pkg/registry/service/ipallocator/allocator.go

```diff
@@ -78,38 +69,34 @@ type Range struct {
 	net *net.IPNet
 	// base is a cached version of the start IP in the CIDR range as a *big.Int
 	base *big.Int
-	// strategy is the strategy for choosing the next available IP out of the range
-	strategy allocateStrategy
 	// max is the maximum size of the usable addresses in the range
 	max int
 
-	// lock guards the following members
-	lock sync.Mutex
-	// count is the number of currently allocated elements in the range
-	count int
-	// allocated is a bit array of the allocated ips in the range
-	allocated *big.Int
+	alloc allocator.Interface
 }
 
-// allocateStrategy is a search strategy in the allocation map for a valid IP.
-type allocateStrategy func(allocated *big.Int, max, count int) (int, error)
-
-// NewCIDRRange creates a Range over a net.IPNet.
-func NewCIDRRange(cidr *net.IPNet) *Range {
+// NewAllocatorCIDRRange creates a Range over a net.IPNet, calling allocatorFactory to construct the backing store.
+func NewAllocatorCIDRRange(cidr *net.IPNet, allocatorFactory allocator.AllocatorFactory) *Range {
 	max := RangeSize(cidr)
 	base := bigForIP(cidr.IP)
-	r := Range{
-		net:      cidr,
-		strategy: randomScanStrategy,
-		base:     base.Add(base, big.NewInt(1)), // don't use the network base
-		max:      maximum(0, int(max-2)),        // don't use the network broadcast
+	rangeSpec := cidr.String()
 
-		allocated: big.NewInt(0),
-		count:     0,
+	r := Range{
+		net:  cidr,
+		base: base.Add(base, big.NewInt(1)), // don't use the network base
+		max:  maximum(0, int(max-2)),        // don't use the network broadcast,
 	}
+	r.alloc = allocatorFactory(r.max, rangeSpec)
 	return &r
 }
 
+// Helper that wraps NewAllocatorCIDRRange, for creating a range backed by an in-memory store.
+func NewCIDRRange(cidr *net.IPNet) *Range {
+	return NewAllocatorCIDRRange(cidr, func(max int, rangeSpec string) allocator.Interface {
```

Comment:
I tried to make something less funky, but I don't think it can be done, because ipallocator.Range calculates the size of the resource pool, so must build the allocator, but sometimes it must build an etcd allocator & sometimes an in-memory allocator.  Suggestions welcome of course :-)